Support for pre-defined profiler templates

jbachorik opened the pull request:
Looking for an initial idea validation here.

The idea is to provide a quick way to select a pre-configured profiler settings such that a customer can switch to 'minimal impact' config without manually creating an override template file etc.

The way to achieve this is to support 'built-in' templates. They are stored as resources in the jar file and can be specified in stead of the override full path. 

Currently supported built-in templates are:
* minimal
* comprehensive

The override templates are containing only overrides - so, if anything changes in dd.jfp and is not overridden it will still be applicable.

The usage is:
* `-Ddd.profiling.jfr-template-override-file=minimal.jfp` for the minimal config
* `-Ddd.profiling.jfr-template-override-file=comprehensive.jfp` for the full config

thegreystone commented:
I'm partial to having separate override files and letting the name simply attempt to load the named jfp file from a well known location (in this case using the agent classloader), just like specifying a simple name for a JFR template (debug / profiling / <myname>) will attempt to find the template file (debug.jfc / profiling.jfc / <myname>.jfc) in a well defined location (jre/lib/jfr). That doesn't restrict us to a certain number of templates and makes it very clear what the resulting overrides will be (no implicit resolution rules depending on the number of entries). Also, adding more templates, or removing templates, will require no code changes.

That said, I agree this syntax is convenient.

jbachorik commented:
So, if we have separate template file per level will it be a full copy of the default one with some extra tweaks for that level?
I was also thinking about 'layering' approach but it complicates the readability and maintainability even further :(

The only thing I found 'difficult' for the template-per-level implementation was not to get lost when updating thresholds in 3 different places :)

mar-kolya commented:
`jdk.JavaMonitorWait#threshold=[500 ms, 100 ms, 10 ms]` - FWIW this seems confusing to me because this depends on how levels are defined in the code and adding more/different levels would be be a pain and potentially vastly incompatible with client's customs setups (unless I'm missing something).

jbachorik commented:
@mar-kolya Yes, I agree. Ideally, we should support maps. But then the complexity will just explode and it becomes easier to add the override template 🤷 
Unfortunately, file-per-level has its own shortcomings as well. It is quite difficult to identify the places which are supposed to be configured differently with the rest staying the same - I know because I tried it as the first pass.

I have also considered splitting the settings by the area they cover (memory, execution, locks etc.) but that also ended up badly when the combinations exploded and, again, creating an override file seemed as a much straight-forward solution :(

But it seems that the generic consensus is forming that it might be less confusing than this extended syntax.

Another possibility is that this is not required at all because our current flow is to take a look at the problematic recording and recommend to change the settings for events with large amounts of instances (and collecting stacktraces). And for this kind of control anything will be more complex than simply providing the override file.

This makes me wonder - wouldn't it be better to rather have analyzer rules popping up warning about too costly events with an exact snippet what to add to the overrides file?

mar-kolya commented:
I'm not sure I see why (at least in theory) 'config inheritance' would not work: when we have bulk of 'general' config in top level file and then just changes for 'mode specific' files.

jbachorik commented:
The inheritance might work - but it might cause surprises when eg. an event is disabled in the default config but the sub-config is just redefining the threshold. In order to prevent this you need to copy all event settings to the sub-config, not just the overrides.
I've had the inheritance based config implemented but since I started with 'profiling areas' instead of plain levels it was quite messy. I can give it another try just with those 3 levels.

jpbempel commented:
The predefined override template seems good to me and easy to use.
However I think the following
```
jdk.ThreadSleep#threshold=[500 ms, 100 ms, 10 ms]
```
is a leftover of the previous implementation, right?

jbachorik commented:
@jpbempel Ah, right! Thanks for noticing!

jpbempel commented on dd-java-agent/agent-profiling/profiling-controller-openjdk/src/main/resources/jfr/dd.jfp:
```diff
@@ -152,9 +152,9 @@ jdk.TenuringDistribution#enabled=true
+jdk.OldObjectSample#enabled=false
 jdk.OldObjectSample#stackTrace=true
-jdk.OldObjectSample#cutoff=0 ms
+jdk.OldObjectSample#cutoff=100 ms
```
What's the rationale for changing this threshold?

jpbempel commented on dd-java-agent/agent-profiling/profiling-controller-openjdk/src/main/resources/jfr/overrides/comprehensive.jfp:
```diff
@@ -0,0 +1,18 @@
+jdk.ThreadSleep#threshold=10 ms
+jdk.ThreadPark#enabled=true
+jdk.ThreadPark#stackTrace=true
+jdk.ThreadPark#threshold=510 ms
```
intended to be 510?
500 or 10?

jbachorik replied:
10. Typing error :/

jbachorik commented:
Ok, now when all tests passed, can I get the second review, plz?

dbenamydd commented:
I only took a quick skim so might have missed something but:

How does this stack with the existing `profiling.jfr-template-override-file`?

> * `-Ddd.profiling.jfr-template-override-file=minimal.jfp` for the minimal config
> * `-Ddd.profiling.jfr-template-override-file=comprehensive.jfp` for the full config

IMO these option names sound like they should be paths. The fact that we've implemented them as such seems like an implementation detail from the user's perspective. Ie I could see someone trying to pass `-Ddd.profiling.jfr-template-override-file=/their/path`. Also, jfr itself is kind of an implementation detail. For example we could imagine a future where we have an alternate collection mechanism, eg for native code, where it might sense to equivalently ease up on data collection. If, from the user's perspective, there are just a couple of specific strings that can work here, and they could coherently refer to multiple collection mechanisms, maybe it should be something like `dd.profiling.collection-level=...`. Also depends on the first question...

jbachorik commented:
Hi @dbenamydd 

> I only took a quick skim so might have missed something but:
>
>Are we getting rid of the previous mechanism for passing in a template file? If so, what happens if someone was using it?
No, we are not getting rid of the previous mechanism.
If someone is using their own override they will keep on using it. These built-in templates are just provided as a convenience so users don't have to create their own overrides for the most typical situations.

>
>-Ddd.profiling.jfr-template-override-file=minimal.jfp for the minimal config
>-Ddd.profiling.jfr-template-override-file=comprehensive.jfp for the full config
>IMO these option names sound like they should be paths. The fact that we've implemented them as such seems like an 
Yes, it is a path. With a convenience shortcut that the JAR resources are searched first for the matching template files.
I decided for this implementation in order not to introduce yet another configuration flag and also because it is consistent - you still define an override template it is just that DataDog provides a bunch of built-in ones you can directly use.

>implementation detail from the user's perspective. Ie I could see someone trying to pass -Ddd.profiling.jfr-template->override-file=/their/path. Also, jfr itself is kind of an implementation detail. For example we could imagine a future where 
Well, this config name was chosen in the beginning and it has been in use since (we have references in troubleshooting, docs, etc.) so it would be better if we keep it - and it affects specifically JFR. If we ever have a different collection mechanism for Java we will have to have different config options to set it up.

>we have an alternate collection mechanism, eg for native code, where it might sense to equivalently ease up on data >collection. If, from the user's perspective, there are just a couple of specific strings that can work here, and they could >coherently refer to multiple collection mechanisms, maybe it should be something like dd.profiling.collection-level=.... 
Well, in fact I started this PR with 'level' based impl. But it was not meat warmly and for the sake of simplicity I reverted back to providing built-in templates.

>Also depends on the first question...

I hope this answered the questions ...

thegreystone commented:
The jfp format is simply a properties version of the jfc format, and it is therefore strongly tied to flight recorder template format. The only reason we're not using jfc files directly, is that we don't need the expressivity of the jfc format (for example, we don't need the possibility to add controls for affecting groups of settings (used by JMC)), and I didn't want to complicate the code unnecessarily with a lot of XML parsing (jfc files are in XML) for our agent. They are also easier to change for a customer. The first jfp files were created through an XSLT from jfc files with our default overrides.

I strongly believe that we should handle these jfp files just like jfc files - it makes the relationship clearer, and it also makes it easy for people knowing jfc files understand how to handle jfp files and vice versa. Principle of least surprise.

This means that we should also allow people to specify only the name part for built in templates - just like jfr supports template=default -> default.jfc searched for in default location, template=profile -> profile.jfc searched in default location.

dbenamydd commented:
Thanks for the answers.

lpriima commented on dd-java-agent/agent-profiling/profiling-controller-openjdk/src/main/java/com/datadog/profiling/controller/openjdk/OpenJdkController.java:
```diff
@@ -22,12 +22,14 @@
  * unmodified for JDK 8+ once JFR has been back-ported. The Oracle JDK implementation will be far
  * messier... ;)
  */
+@Slf4j
```
forget to delete ?

jbachorik replied:
Right. Will remove.

lpriima commented on dd-java-agent/agent-profiling/profiling-controller-openjdk/src/test/java/com/datadog/profiling/controller/openjdk/JfpUtilsTest.java:
```diff
@@ -29,4 +40,22 @@ public void testLoadingContinuousConfigWithOverride() throws IOException {
   }
+
+  @ParameterizedTest
+  @ValueSource(strings = {"minimal", "minimal.jfp"})
```
I can suggest to test `{"minimal", "minimal"}` to cover path where you add `.jfp`

jbachorik replied:
This is actually testing that - the parameter values are there exactly to check both fully specified and with no extension.

lpriima commented on dd-java-agent/agent-profiling/profiling-controller-openjdk/src/main/java/com/datadog/profiling/controller/openjdk/JfpUtils.java:
```diff
@@ -53,15 +57,37 @@ private static InputStream getNamedResource(final String name) {
+    if (overridesFileName != null) {
+      InputStream overrideStream = null;
+      try {
+        if (!overridesFileName.toLowerCase().endsWith(JFP_EXTENSION)) {
+          overridesFileName = overridesFileName + JFP_EXTENSION;
+        }
+        File override = new File(overridesFileName);
```
These lines don't deal with closable resources, so they can be moved outside/before try {} block and maybe use try-with:

```
      if (!overridesFileName.toLowerCase().endsWith(JFP_EXTENSION)) {
        overridesFileName = overridesFileName + JFP_EXTENSION;
      }
      File override = new File(overridesFileName);
      try (InputStream overrideStream =
          override.exists()
              ? new FileInputStream(override)
              : getNamedResource(OVERRIDES_PATH + overridesFileName)) {
        result.putAll(readJfpFile(overrideStream));  // `readJfpFile` already has a null check
      }
```

jbachorik replied:
Yep. It can be simplified. Will take a look.
